lib/discourse_js_processor.rb

```diff
@@ -100,19 +103,21 @@ def self.skip_module?(data)
   class Transpiler
     @mutex = Mutex.new
     @ctx_init = Mutex.new
+    @processor_mutex = Mutex.new
 
     def self.mutex
       @mutex
     end
 
-    def self.load_file_in_context(ctx, path, wrap_in_module: nil)
-      contents = File.read("#{Rails.root}/app/assets/javascripts/#{path}")
-      contents = <<~JS if wrap_in_module
-          define(#{wrap_in_module.to_json}, ["exports", "require", "module"], function(exports, require, module){
-            #{contents}
-          });
-        JS
-      ctx.eval(contents, filename: path)
+    def self.generate_js_processor
+      @processor_mutex.synchronize do
+        if Rails.env.development? || Rails.env.test?
+          !File.exist?("#{Rails.root}/app/assets/javascripts/compiled-js-processor.js")
```

Comment:
What happens when we want to make changes to any of the source files? I think with the current implementation we'd have to delete the `tmp/js-processor/...` file manually so that it's regenerated?

I think ideally the name of the file would include some kind of hash/timestamp. Maybe simplest option is 'most recent modified time in `app/assets/javascripts`' ? That's kinda how we manage things [for stylesheets](https://github.com/discourse/discourse/blob/f2446cd0db270ab0449352a329f50bf5368f0b8d/lib/stylesheet/manager.rb#L177-L180)

Reply:
In dev and test envs it recompiles it every time `generate_js_processor` is called, so no need to delete any files. I also just added an entry to `config/initializers/000-development_reload_warnings.rb` so that any change to `js-processor.js` will trigger server reload in dev 😃